Identity lenses using `typed`

This code fails:

```haskell
{-# LANGUAGE FlexibleContexts, DeriveGeneric, TypeApplications #-}

module GenLens where

import GHC.Generics
import Control.Lens
import Data.Generics.Product

data Foobar = Foobar Int Char String
    deriving Generic

foo :: HasType Foobar ctx => ctx -> Char
foo ctx = case view (typed @Foobar) ctx of
    Foobar _ c _ -> c

bar :: IO ()
bar = do
    print (foo (Foobar 3 'a' "Hello"))
```

with this error message:

```
/home/matt/Projects/playground/src/GenLens.hs:18:12: error:
    • The type Foobar does not contain a value of type Foobar
    • In the first argument of ‘print’, namely
        ‘(foo (Foobar 3 'a' "Hello"))’
      In a stmt of a 'do' block: print (foo (Foobar 3 'a' "Hello"))
      In the expression: do print (foo (Foobar 3 'a' "Hello"))
   |
18 |     print (foo (Foobar 3 'a' "Hello"))
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I would expect that it would work out to the identity lens: `id :: Lens' a a`

hmm, I tried changing `CollectTotalType` to:

```haskell
type family CollectTotalType t f :: TypeStat where
  CollectTotalType t (C1 ('MetaCons ctor _ _) f)
    = AddToStat ctor (CountType t f) EmptyStat
  CollectTotalType t (M1 D m a)
    = CheckEqual t (Rep t) (M1 D m a)
  CollectTotalType t (M1 _ _ r)
    = CollectTotalType t r
  CollectTotalType t (l :+: r)
    = MergeStat (CollectTotalType t l) (CollectTotalType t r)

type family CheckEqual o a b where
  CheckEqual _ (M1 D ('MetaData n a b c) d) (M1 D ('MetaData n a b c) d)
    = AddToStat n 'One EmptyStat
  CheckEqual o _ (M1 _ _ r)
    = CollectTotalType o r
```

but now the error I get is hard to decipher:

```
/home/matt/Projects/generic-lens/test/Test71.hs:17:7: error:
    • Couldn't match type ‘[Char]’ with ‘Foobar’
        arising from a functional dependency between:
          constraint ‘generic-lens-1.1.0.0:Data.Generics.Product.Internal.GLens.GLens
                        (Data.Generics.Product.Typed.HasTotalTypePSym Foobar)
                        (K1 R String)
                        (K1 R String)
                        Foobar
                        Foobar’
            arising from a use of ‘foo’
          instance ‘generic-lens-1.1.0.0:Data.Generics.Product.Internal.GLens.GLens
                      pred (K1 r a) (K1 r b) a b’
            at <no location info>
    • In the expression: foo (Foobar 3 'a' "Hello")
      In an equation for ‘bar’: bar = foo (Foobar 3 'a' "Hello")
   |
17 | bar = foo (Foobar 3 'a' "Hello")
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

--- a/src/Data/Generics/Product/Typed.hs
+++ b/src/Data/Generics/Product/Typed.hs
@@ -117,8 +117,11 @@ instance
   typed f s = VL.ravel (repLens . glens @(HasTotalTypePSym a)) f s
 
 instance {-# OVERLAPPING #-} HasType a a where
-    typed = id
-    {-# INLINE typed #-}
+    getTyped = id
+    {-# INLINE getTyped #-}
+
+    setTyped a _ = a
+    {-# INLINE setTyped #-}
 
 -- See Note [Uncluttering type signatures]
 instance {-# OVERLAPPING #-} HasType a Void where

--- a/src/Data/Generics/Product/Typed.hs
+++ b/src/Data/Generics/Product/Typed.hs
@@ -116,6 +116,10 @@ instance
 
   typed f s = VL.ravel (repLens . glens @(HasTotalTypePSym a)) f s
 
+instance {-# OVERLAPPING #-} HasType a a where
+    typed = id
+    {-# INLINE typed #-}
+
 -- See Note [Uncluttering type signatures]
 instance {-# OVERLAPPING #-} HasType a Void where
   typed = undefined

--- a/test/Test71.hs
+++ b/test/Test71.hs
@@ -1,0 +1,17 @@
+{-# LANGUAGE FlexibleContexts, DeriveGeneric, TypeApplications #-}
+
+module Test71 where
+
+import GHC.Generics
+import Control.Lens
+import Data.Generics.Product
+
+data Foobar = Foobar Int Char String
+    deriving Generic
+
+foo :: HasType Foobar ctx => ctx -> Char
+foo ctx = case view (typed @Foobar) ctx of
+    Foobar _ c _ -> c
+
+bar :: Char
+bar = foo (Foobar 3 'a' "Hello")

